Add WebRender integration to Servo.

glennw opened the pull request:
WebRender is an experimental GPU accelerated rendering backend for Servo.

The WebRender backend can be specified by running Servo with the -w option (otherwise the default rendering backend will be used).

WebRender has many bugs, and missing features - but it is usable to browse most websites - please report any WebRender specific rendering bugs you encounter!

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9589)

<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @ecoal95: components/script/dom/webglrenderingcontext.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

glennw commented:
Since this is such a large patch, if anyone can take a look at their areas and do partial reviews, that would be great!

ecoal95 commented:
Awesome work!

Reviewed everything (except the glsl files) and I couldn't spot anything that looked wrong or deserved a change... Not an expert in every part of the codebase this patch touches though.

Left a few notes for myself and for the curious regarding the moved WebGL code.

---

Review status: 0 of 76 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/canvas/webgl_paint_thread.rs, line 182 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9589#-KA7cGNifpk4EAa8gq3L:-KA7cGNifpk4EAa8gq3M:1117578804) ([raw file](https://github.com/servo/servo/blob/cb8be1434f61e6bfe4639d42b1422329b0451a47/components/canvas/webgl_paint_thread.rs#L182)):*
If someone's curious, all this logic is moved to webrender_traits in order to be shared with the webrender WebGL backend, which is different from Servo's.

---

*[components/canvas_traits/lib.rs, line 43 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9589#-KA7bsmcexAUnUISdNuR:-KA7bsmcexAUnUISdNuS:261052928) ([raw file](https://github.com/servo/servo/blob/def3c2b7dab93fa50fa9bb51df23769bbeb2343b/components/canvas_traits/lib.rs#L43)):*
Just as a note to self when this merges: An E-Easy should be filled as a follow-up to remove this re-export(s), which was done to avoid extra rebase pain.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9589#-:-KA7eydn-zQlesvyRxVt:-1089853933)*
<!-- Sent from Reviewable.io -->

pcwalton commented:
Should I review parts of this or am I too close to being the author for that?

jgraham commented:
I don't know if it's needed, but note that wptrunner supports switching the expected results of tests based on the render backend, so if we plan to run webrender in CI we should check in metadata updates and not disable tests.

RoxasShadow commented on resources/filter.vs.glsl:
```diff
@@ -0,0 +1,11 @@
+{
+	vColorTexCoord = aColorTexCoordRectTop.xy;
+	vMaskTexCoord = aMaskTexCoordRectTop.xy;
+    gl_Position = uTransform * vec4(aPosition, 1.0);
```
The alignment is broken in this line

RoxasShadow commented on resources/quad.vs.glsl:
```diff
@@ -0,0 +1,122 @@
+        if (aPosition.x == 0.0) {
+            localPos.x = rect_origin.x;
+            if (isBorderCorner) {
+                vColor = !isBottomTriangle ? aColorRectTR : aColorRectBL;
```
Shouldn't `vColor = isBottomTriangle ? aColorRectBL : aColorRectTR;` work as well?

RoxasShadow commented on resources/tile.vs.glsl:
```diff
@@ -0,0 +1,11 @@
+
+void main(void)
+{
+	vColorTexCoord = aBorderRadii.xy;
```
Is this an unintended tab?

jdm commented:
```
   Compiling gfx_tests v0.0.1 (file:///home/travis/build/servo/servo/components/servo)

/home/travis/build/servo/servo/tests/unit/gfx/font_cache_thread.rs:14:27: 14:57 error: this function takes 2 parameters but 1 parameter was supplied [E0061]

/home/travis/build/servo/servo/tests/unit/gfx/font_cache_thread.rs:14   let font_cache_thread = FontCacheThread::new(inp_chan);

                                                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/home/travis/build/servo/servo/tests/unit/gfx/font_cache_thread.rs:14:27: 14:57 help: run `rustc --explain E0061` to see a detailed explanation

error: aborting due to previous error

Build failed, waiting for other jobs to finish...

Could not compile `gfx_tests`.
```

steveklabnik commented:
Builds on my machine, seems to work just fine :+1: 

There's a lot of strange indentation in the GSL files, not sure if that's on purpose or not.

glennw commented:
Fixed up the unit tests and the notes about the shaders.

glennw commented:
@jgraham Great! Are there any docs / examples on how I should set up the files to have different expectations?

jgraham commented:
You can use `backend == "cpu"` or `backend == "webrender"` in the expectation files in the same way you can use e.g. `os == "mac"`. If you create a log file from a non-webrender run and a log file from a webrender run then use `mach update-wpt default.log webrender.log` then it will add the necessary conditions for any tests with different results.

mrobinson commented on components/gfx/display_list/mod.rs:
```diff
@@ -976,6 +985,7 @@ pub enum DisplayItem {
     StackingContextClass(Arc<StackingContext>),
     LayeredItemClass(Box<LayeredItem>),
     NoopClass(Box<BaseDisplayItem>),
+    IframeClass(Box<IframeDisplayItem>),
```
It feels like NoopClass could be removed in favor of the new IframeClass. NoopClass exits just as a placeholder so that iframes can hold a place in the display list without drawing anything.

glennw replied:
@mrobinson Sounds good - is it fine to this as a follow up once this PR lands do you think?

mrobinson replied:
@glennw Yeah, I think it would be fine to do in a separate patch.

ecoal95 commented:
Review status: 0 of 77 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/gfx/font_cache_thread.rs, line 300 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9589#-KADIMf6iZqqTeLkT07A:-KADIMf6iZqqTeLkT07B:369518719) ([raw file](https://github.com/servo/servo/blob/6e3a24cfb74f0b241344dc91825e334b4eb0a569/components/gfx/font_cache_thread.rs#L300)):*
This could probably use [or_insert_with](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with). Something like:

```rust
let font_key = self.webrender_api.as_ref().map(|webrender_api| {
    self.webrender_fonts.entry(template.identifier.clone()).or_insert_with(|| {
        match (template.bytes_if_in_memory(), template.native_font()) {
            (Some(bytes), _) => webrender_api.add_raw_font(bytes),
            (None, Some(native_font)) => webrender_api.add_native_font(native_font),
            (None, None) => webrender_api.add_raw_font(template.bytes().clone()),
        }
    })
});
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9589)*
<!-- Sent from Reviewable.io -->

mbrubeck commented:
Reviewed 3 of 76 files at r1.
Review status: 3 of 77 files reviewed at latest revision, 8 unresolved discussions.

---

*[components/compositing/compositor.rs, line 1422 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9589#-KAGIcK7rIP2GkajcdLO:-KAGIcK7rIP2GkajcdLP:-368944296) ([raw file](https://github.com/servo/servo/blob/6e3a24cfb74f0b241344dc91825e334b4eb0a569/components/compositing/compositor.rs#L1422)):*
Should have a TODO comment and/or follow-up issue to implement zooming.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9589)*
<!-- Sent from Reviewable.io -->

mbrubeck commented:
Reviewed 7 of 76 files at r1, 1 of 1 files at r2.
Review status: 11 of 77 files reviewed at latest revision, 9 unresolved discussions.

---

*[tests/wpt/harness/wptrunner/browsers/servo.py, line 61 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9589#-KAHNr95fM46vFVIU3wb:-KAHNr98AerptKquv8YI:-526050566) ([raw file](https://github.com/servo/servo/blob/6e3a24cfb74f0b241344dc91825e334b4eb0a569/tests/wpt/harness/wptrunner/browsers/servo.py#L61)):*
This will also need to be upstreamed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9589)*
<!-- Sent from Reviewable.io -->

mbrubeck commented:
Reviewed 4 of 76 files at r1.
Review status: 15 of 77 files reviewed at latest revision, 11 unresolved discussions.

---

*[components/util/Cargo.toml, line 29 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9589#-KAHOu3T0LmAwWIxbQwL:-KAHOu3UlZ5BfjNJLUeL:-875459903) ([raw file](https://github.com/servo/servo/blob/6e3a24cfb74f0b241344dc91825e334b4eb0a569/components/util/Cargo.toml#L29)):*
This appears to be unused.

---

*[components/util/lib.rs, line 49 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9589#-KAHOxuYy5fCxlEiKz6N:-KAHOxuYy5fCxlEiKz6O:-2030536049) ([raw file](https://github.com/servo/servo/blob/6e3a24cfb74f0b241344dc91825e334b4eb0a569/components/util/lib.rs#L49)):*
This looks unused.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9589)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #9622) made this pull request unmergeable. Please resolve the merge conflicts.

pcwalton commented on components/compositing/compositor.rs:
```diff
@@ -261,7 +269,9 @@ pub enum CompositeTarget {
-fn initialize_png(width: usize, height: usize) -> (Vec<gl::GLuint>, Vec<gl::GLuint>) {
+fn initialize_png(width: usize, height: usize) -> (Vec<gl::GLuint>,
+                                                   Vec<gl::GLuint>,
+                                                   Vec<gl::GLuint>) {
```
Could we change the return value of this to be a struct?

pcwalton commented on components/compositing/compositor.rs:
```diff
@@ -1172,6 +1249,32 @@ impl<Window: WindowMethods> IOCompositor<Window> {
+
+        if let Some(ref webrender_api) = self.webrender_api {
+            if let Some(root_pipeline_id) = self.get_root_pipeline_id() {
+                if let Some(root_pipeline) = self.pipeline(root_pipeline_id) {
```
Maybe something like:

    let root_pipeline_id = match self.get_root_pipeline_id() {
        Some(root_pipeline_id) => root_pipeline_id,
        None => return,
    };
    let root_pipeline = match self.pipeline(root_pipeline_id) {
        Some(root_pipeline) => root_pipeline,
        None => return,
    };

Just to reduce rightward drift.

pcwalton commented on components/compositing/compositor.rs:
```diff
@@ -1184,6 +1287,21 @@ impl<Window: WindowMethods> IOCompositor<Window> {
 
+        if let Some(ref webrender_api) = self.webrender_api {
+            if let Some(root_pipeline_id) = self.get_root_pipeline_id() {
+                if let Some(root_pipeline) = self.pipeline(root_pipeline_id) {
```
Ditto.

pcwalton commented on components/compositing/compositor.rs:
```diff
@@ -1285,30 +1403,52 @@ impl<Window: WindowMethods> IOCompositor<Window> {
+                // Batch up all scroll events into one, or else we'll do way too much painting.
+                let mut total_delta = None;
+                let mut last_cursor = Point2D::zero();
+                for scroll_event in std_mem::replace(&mut self.pending_scroll_zoom_events, vec![]) {
```
Can we use `for scroll_event in self.pending_scroll_zoom_events.drain(..) {` now?

pcwalton commented on components/compositing/compositor.rs:
```diff
@@ -1656,7 +1800,19 @@ impl<Window: WindowMethods> IOCompositor<Window> {
+                                let epoch = Epoch(epoch);
+                                pipeline_epochs.insert(*id, epoch);
+                            }
+                            None => {}
```
nit: `if let`

pcwalton commented on components/compositing/compositor.rs:
```diff
@@ -1656,7 +1800,19 @@ impl<Window: WindowMethods> IOCompositor<Window> {
+                        let webrender_pipeline_id = id.to_webrender();
+                        match webrender.current_epoch(webrender_pipeline_id) {
+                            Some(epoch) => {
+                                let webrender_traits::Epoch(epoch) = epoch;
```
Could fold this `let` into the `Some` above: `Some(webrender_traits::Epoch(epoch)) => {`

pcwalton commented on components/layout/layout_thread.rs:
```diff
@@ -908,9 +917,40 @@ impl LayoutThread {
+                    let epoch = webrender_traits::Epoch(epoch_number);
+                    let pipeline_id = self.id.to_webrender();
+
+                    // TODO(gw) For now only create a root scrolling layer!
```
What are the ramifications of this? Do iframes not work?

glennw replied:
Iframes work - there is a root scrolling layer per pipeline. What doesn't work yet is overflow: scroll within a normal page - I was going to wait until mrobinson's display list changes for that, since we have a heap of hacks at the moment related to creating a fake stacking context to handle this case in Servo.

pcwalton commented:
I scanned through this patch and it looks good now. I wonder if we should put the shaders inside a subdirectory of `resources/` to avoid cluttering it up, perhaps `resources/shaders/` or something?

glennw commented:
@pcwalton Sounds good, I'll move the shaders now.

Ms2ger commented:
This needs squashing before landed.

gw3583 committed:
Add WebRender integration to Servo.

WebRender is an experimental GPU accelerated rendering backend for Servo.

The WebRender backend can be specified by running Servo with the -w option (otherwise the default rendering backend will be used).

WebRender has many bugs, and missing features - but it is usable to browse most websites - please report any WebRender specific rendering bugs you encounter!

glennw commented:
Squashed and rebased.

pcwalton commented:
r=me, btw.

mbrubeck commented:
Reviewed 6 of 104 files at r1, 1 of 15 files at r6, 8 of 21 files at r9.
Review status: 21 of 75 files reviewed at latest revision, 19 unresolved discussions.

---

*[components/gfx/platform/freetype/font_template.rs, line 42 \[r11\]](https://reviewable.io:443/reviews/servo/servo/9589#-KAlGrAm2YmM2HMnhbM5:-KAlGrAm2YmM2HMnhbM6:987366233) ([raw file](https://github.com/servo/servo/blob/a86ba3a0c5759d89f0595e5caf43aa9284f3c7a8/components/gfx/platform/freetype/font_template.rs#L42)):*
Currently this comment is incorrect. The disk I/O is already done in `FontTemplateData::new`. Should there be a TODO comment or follow-up issue about this?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9589)*
<!-- Sent from Reviewable.io -->

glennw commented:
@mbrubeck I updated that comment to note that it *might* do disk I/O depending on the platform (it does on Mac, but not on Linux).

glennw commented:
r=pcwalton

glennw commented:
@bors-servo r=pcwalton

bors-servo commented:
:pushpin: Commit 3991d4e has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 3991d4e240ef723b091bb6b0c0e19da295fe33a5 -->

bors-servo commented:
:hourglass: Testing commit 3991d4e with merge 15e77a8...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/1730)

glennw commented:
Fixed the android build failure in WR, and updated the cargo.lock files.

glennw commented:
r=pcwalton

glennw commented:
@bors-servo r=pcwalton

bors-servo commented:
:pushpin: Commit c0531c3 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton c0531c312fdb0783e4d121b4c2d7f15d4f5cdc1f -->

bors-servo commented:
:hourglass: Testing commit c0531c3 with merge 3640b82...

samlh commented:
> 
>   ▶ PASS [expected FAIL] /css21_dev/html4/floats-wrap-bfc-002-right-table.htm
> 
>   ▶ PASS [expected FAIL] /css21_dev/html4/position-relative-035.htm

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1763)

KiChjang commented:
```
Ran 9335 tests finished in 1478.0 seconds.
  • 9333 ran as expected. 58 tests skipped.
  • 2 tests passed unexpectedly

Tests with unexpected results:
  ▶ PASS [expected FAIL] /css-transforms-1_dev/html/transform-input-017.htm

  ▶ PASS [expected FAIL] /css-transforms-1_dev/html/transform-input-018.htm
```
Looking really good!

jdm commented:
Those are the same tests that are being reported as unexpected passes in #9668. What's going on here?

glennw commented:
@jdm This patch includes the commit in #9668 - I was trying to upstream that individually before this patch got reviewed.

glennw commented:
@bors-servo retry

* #9668 has landed now.

bors-servo commented:
:hourglass: Testing commit c0531c3 with merge ab07b06...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1712), [gonk](http://build.servo.org/builders/gonk/builds/1747), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1720), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1966), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/421), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1778), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1818)
